reps/2023-02-01-wasm-on-ray.md

```diff
@@ -0,0 +1,270 @@
+
+### Other Languages Support
+
+#### Rust
+It is also possible to support other languages that can be compiled to WASM. For example in Rust, you can declare the Ray APIs you want to import in the following way.
+
+```rust
+#[link(wasm_import_module = "ray")] 
+extern "C" {
+  #[link_name="call"]
+  fn rcall(ptr: *const u8, ...) -> u32;
+}
+
+```
+
+Then you can use the `rcall` function to call the remote function. The `ptr` is the function pointer. The following arguments are the arguments of the remote function.
+
+
+#### Other Languages
+
+To use Ray APIs in JavaScript, the Ray APIs need to be exported to the Javascript runtime environment. For example, if the user is using SpiderMonkey, the Ray APIs can be exported to SpiderMonkey and then used in JavaScript.
+
+Futhermore, the new Component Model standard is supposed to be ready very soon. It will provide a standard way to import and export components. We can use the new Component Model standard to declare Ray APIs in WASM so that it will be easier to use Ray APIs in different languages.
+
+
+## Pros and Cons
+Compared with directly supporting other languages, WASM support in Ray has the following pros and cons.
+
+### Pros:
+1. It is easy to support new languages. For example, if we want to support a new language, we only need to compile the language to WASM and then use the WASM support in Ray to run the WASM binary.
+2. Offer the ability to run Ray-As-A-Service. With WASM enabled in Ray, it is possible for users to use untrusted third-party WASM binaries in Ray. As WASM supports a sandboxed environment, the activities of the untrusted third-party WASM binaries can do is limited.
+3. Lightweight Workers. WASM is only 10% the size of Python for comparable tasks. This means that the WASM worker is much smaller than the Python worker. WASM is lightweight and fast to load. It is possible to load the WASM binary in a few milliseconds. Compared to Python, it is much faster to load the WASM binary.
+4. Architecture Agnostic. WASM provides an abstract layer on top of the hardware. It is possible to support different hardware architectures in the same WASM binary.
+5. Easy to support different versions of the same language. For example, if we want to support different versions of JavaScript, we only need to compile different versions of JavaScript to WASM and then use the WASM support in Ray to run the WASM binary.
+
+### Cons:
+1. It is not easy to support all Ray APIs in WASM. For example, there is a concept of `Actor` in Ray. `Actor` is a special type of task that can maintain state. However, WASM is a collection of functions, a stack, and a heap. It is not a simple task to map the concept of `Actor` to WASM. Therefore, we will not support `Actor` in WASM in the first version of WASM support in Ray.
```

Comment:
@iycheng @wilsonwang371 
From your design, if we don't re-design a better lower level API for developers, it's not easy to implement WASM actor, right?
I'm concerning that it's verbose to do all these stuffs in one.
We should separate this REP to 2 proposals: a better lower level API for multi-languages workers proposal + WASM worker proposal.

The benefits are significant:
1. For the lower level API proposal, I guess there're many committers and reviewers want to review, while that the WASM worker proposal is focusing on how we port WASM on Ray, which I believe less of reviewers is available to review due to it requires more knowledge about WASM(For example, I'm lack of that).
2. Different goals should use different REPs for better track in the future.

Reply:
I think WASM workers will not require a better lower-level API to be there in the first place. But our Bytedance team is definitely willing to work together on the lower-level API design & implementation.

Reply:
Besides, the lower-level API is not a part of this REP.

Reply:
I see your point that this REP doesn't include the lower level API here.

I'm worrying about whether there might be a risk, like once this REP getting merged, and then it's impossible to support WASM actor.

If you guys are confident for this, let us make it clearer.

Reply:
there will be ways to support actor in wasm. But we need to choose a solution that is acceptable to most people. That's why this is not a part of the MVP version

Reply:
OK. Please let me know once you have design or some stuffs. Looking forward the contributions.

Reply:
I think the issues here are that no one knows how to do it :(

Before starting actor support, we probably should get to the api layer first and we'll feel comfortable about this at that stage.